src/client/redirect-watch.js

```diff
@@ -8,69 +8,84 @@ import domProcessor from './dom-processor/index';
 
-        this.codeInstrumentation = codeInstrumentation;
-        this.eventSandbox        = eventSandbox;
+        this.lastLocationValue = window.location.toString();
+
+        this._locationWatch(codeInstrumentation);
+        this._linkWatch(listeners);
+        this._formWatch(elementSandbox);
     }
 
-    init () {
-        this._locationWatch(this.codeInstrumentation);
-        this._linkClickWatch(this.eventSandbox);
+    _formWatch (elementSandbox) {
+        elementSandbox.on(elementSandbox.BEFORE_FORM_SUBMIT, e => {
+            var targetWindow = this._getTargetWindow(e.form);
+
+            this._redirectWindow(targetWindow, e.form.action);
+        });
     }
 
-    redirect (url) {
-        this.emit(this.DETECTED_EVENT, parseProxyUrl(url).destUrl);
+    _redirectWindow (window, url) {
+        try {
+            window['%hammerhead%'].redirectWatch.redirect(url);
+        }
+            /*eslint-disable no-empty */
+        catch (e) {
+        }
+        /*eslint-enable no-empty */
     }
 
-    _linkClickWatch (eventSandbox) {
-        eventSandbox.listeners.addInternalEventListener(window, ['click'], e => {
+    _getTargetWindow (el) {
+        var targetWindow = window;
+        var target       = el.getAttribute(domProcessor.getStoredAttrName('target')) ||
+                           el.getAttribute('target');
+
+        if (target) {
+            if (!ElementSandbox._isKeywordTarget(target))
+                targetWindow = windowsStorage.findByName(target) || window;
+            else if (target === '_top')
+                targetWindow = window.top;
+            else if (target === '_parent')
+                targetWindow = window.parent;
+        }
+
+        return targetWindow;
+    }
+
+    _linkWatch (listeners) {
+        listeners.initElementListening(window, ['click']);
+        listeners.addInternalEventListener(window, ['click'], e => {
             var link = e.target;
 
             if (link.tagName && link.tagName.toLowerCase() === 'a' && !isShadowUIElement(link)) {
-                var target       = link.getAttribute(domProcessor.getStoredAttrName('target')) ||
-                                   link.getAttribute('target');
-                var targetWindow = window;
-
-                if (target) {
-                    if (!ElementSandbox._isKeywordTarget(target))
-                        targetWindow = windowsStorage.findByName(target) || window;
-                    else if (target === '_top')
-                        targetWindow = window.top;
-                    else if (target === '_parent')
-                        targetWindow = window.parent;
-                }
-                try {
-                    targetWindow['%hammerhead%'].redirectWatch.redirect(link.href);
-                }
-                    /*eslint-disable no-empty */
-                catch (ex) {
-                }
-                /*eslint-enable no-empty */
+                var targetWindow = this._getTargetWindow(link);
```

Comment:
write `domUtils.isAnchorElement`
intead of `link.tagName && link.tagName.toLowerCase() === 'a'`